balloon/src/lib.rs

```diff
@@ -0,0 +1,347 @@
+};
+use core::convert::{TryFrom, TryInto};
+use core::marker::PhantomData;
+use digest::generic_array::GenericArray;
+use digest::Digest;
+#[cfg(feature = "password-hash")]
+#[cfg_attr(docsrs, doc(cfg(feature = "password-hash")))]
+pub use password_hash::{self, PasswordHash, PasswordHasher, PasswordVerifier};
+#[cfg(all(feature = "alloc", feature = "password-hash"))]
+use password_hash::{Decimal, Ident, ParamsString, Salt};
+
+/// Balloon context.
+///
+/// This is the primary type of this crate's API, and contains the following:
+///
+/// - Default set of [`Params`] to be used
+/// - (Optional) Secret key a.k.a. "pepper" to be used
+#[derive(Clone, Default)]
+pub struct Balloon<'key, D: Digest> {
+    /// Storing which hash function is used
+    pub digest: PhantomData<D>,
+    /// Algorithm parameters
+    pub params: Params,
+    /// Key array
+    pub secret: Option<&'key [u8]>,
+}
+
+impl<'key, D: Digest> Balloon<'key, D> {
+    /// Create a new Balloon context.
+    pub fn new(params: Params, secret: Option<&'key [u8]>) -> Self {
+        Self {
+            digest: PhantomData,
+            params,
+            secret,
+        }
+    }
+
+    /// Hash a password and associated parameters.
+    #[cfg(feature = "alloc")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+    pub fn hash(&self, pwd: &[u8], salt: &[u8]) -> Result<GenericArray<u8, D::OutputSize>> {
+        if self.params.p_cost.get() == 1 {
+            let mut memory = alloc::vec![GenericArray::default(); usize::try_from(self.params.s_cost.get()).unwrap()];
+            self.hash_with_memory(pwd, salt, &mut memory)
+        } else {
+            #[cfg(not(feature = "parallel"))]
+            return Err(Error::ThreadsTooMany);
+            #[cfg(feature = "parallel")]
+            {
+                use core::num::NonZeroU32;
+                use rayon::iter::{IntoParallelIterator, ParallelIterator};
+
+                let mut params = self.params;
+                params.p_cost = NonZeroU32::new(1).unwrap();
+
+                (1..=self.params.p_cost.get())
+                    .into_par_iter()
+                    .map_with((params, self.secret), |(params, secret), index| {
+                        const U32_SIZE: usize = core::mem::size_of::<u32>();
+
+                        // expand salt if necessary
+                        let mut salt = if salt.len() < U32_SIZE {
+                            let mut new_salt = [0; U32_SIZE];
+                            new_salt[..salt.len()].copy_from_slice(salt);
+                            new_salt.to_vec()
+                        } else {
+                            salt.to_vec()
+                        };
+
+                        // add thread index to salt
+                        let byte = salt[..U32_SIZE].try_into().unwrap();
+                        salt[..U32_SIZE].copy_from_slice(
+                            &(u32::from_le_bytes(byte).wrapping_add(index)).to_le_bytes(),
+                        );
```

Comment:
This is what the paper proposes:
> BalloonM(p,s) := Balloon(p, s‖“1”) ⊕···⊕Balloon(p, s‖“M”)`

I don't know what `||` is, I found this: https://en.wikipedia.org/wiki/Parallel_(operator).
The reference implementation does this: [libballoon/balloon.c#L62-L76](https://github.com/henrycg/balloon/blob/0ae35a8879a97f4c7c44ee7fe77ea6522c607ef7/libballoon/balloon.c#L62-L76), but I didn't want to blindly translate some code.

Reply:
|| is the concatenation operation

Reply:
I see, that could simply allow me to just add it to the hash, no reason to allocate.

Reply:
Done.